core/client/fs/src/main/java/alluxio/master/MasterHealthCheckClient.java

```diff
@@ -0,0 +1,103 @@
+ */
+
+package alluxio.master;
+
+import alluxio.AlluxioURI;
+import alluxio.HealthCheckClient;
+import alluxio.client.file.FileSystem;
+import alluxio.util.ShellUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetSocketAddress;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * MasterHealthCheckClient check whether Alluxio master is serving RPC request and
+ * if the AlluxioMaster process is running in all master hosts.
+ */
+public class MasterHealthCheckClient implements HealthCheckClient {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MasterHealthCheckClient.class);
+
+  private Thread mMasterServingThread;
+  private Thread mMastersProcessCheckThread;
+  private ScheduledExecutorService mExecutorService = Executors.newScheduledThreadPool(2);
+
+  /**
+   * Creates a master health check client.
+   */
+  public MasterHealthCheckClient() {
+    mMasterServingThread = new Thread(() -> {
+      try {
+        LOG.debug("Checking master is serving requests");
+        FileSystem fs = FileSystem.Factory.get();
+        fs.exists(new AlluxioURI("/"));
+        LOG.debug("Master is serving requests");
+      } catch (Throwable e) {
+        throw new RuntimeException(e);
+      }
+    }, "MasterServingThread");
+
+    mMastersProcessCheckThread = new Thread(() -> {
+      MasterInquireClient client = MasterInquireClient.Factory.create();
+      try {
+        List<InetSocketAddress> addresses = client.getMasterRpcAddresses();
+        for (InetSocketAddress address : addresses) {
+          String host = address.getHostName();
+          LOG.debug("Master health check on node {}", host);
+          String cmd = String.format("ssh %s %s %s", ShellUtils.COMMON_SSH_OPTS, host,
+                  "ps -ef | "
+                  + "grep -v \"alluxio.master.AlluxioMasterMonitor\" | "
+                  + "grep \"alluxio.master.AlluxioMaster\" | "
+                  + "grep \"java\" | "
+                  + "awk '{ print $2; }'");
+          LOG.debug("Executing: {}", cmd);
+          String output = ShellUtils.execCommand("bash", "-c", cmd);
+          if (output.isEmpty()) {
+            throw new IllegalStateException(
+                    String.format("Master process is not running on the host %s", host));
+          }
+          LOG.debug("Master running on node {} with pid={}", host, output);
+        }
+      } catch (Throwable e) {
+        if (!mExecutorService.isShutdown()) {
+          mExecutorService.shutdownNow();
+        }
+        throw new RuntimeException(e);
+      }
+    }, "MastersProcessCheckThread");
+  }
+
+  @Override
+  public boolean isServing() {
+    try {
+      mExecutorService.scheduleAtFixedRate(mMastersProcessCheckThread, 0, 5,
```

Comment:
how does this report the error if a non-primary master goes down?

Reply:
At the moment I am assuming that if one of the master hosts is not running  the ```AlluxioMaster``` process (regardless if it's primary or not)  the cluster is in an invalid state, thus this check will kill the ```ExecutorService``` and cause the ```MasterServingThread``` to fail.

Reply:
Oh sorry, I missed that the thread was shutting down the executor service running it. This is kind of a strange behavior, since usually an executor service is not shut down from a thread running inside it. The ExecutorService manages the Thread, but the Thread also manages the ExecutorService. I'd like to avoid the circular pattern if possible. We could do something like poll both threads from the main thread in a loop until either the process check thread fails or the RPC thread succeeds. That would be easier to implement if the process thread did its own scheduling so that it runs only once.

Reply:
@aaudiber Yes agree, I ran into some problems where I couldn't interrupt the call to ```fs.exist``` needed to wait for the default retries (around a 100 tries), which was really time consuming. Do you have a any suggestion how to deal with this kind of issue? seems that either I need to adjust manually the number of retries or I need to add support for interrupt to the ```fs.exist``` method.

Reply:
It's difficult to make a complex thread reliably interruptible. Even if we don't do anything wrong in our own code, there could be places in the Thrift client or other libraries that don't properly support interrupts. The code will be more stable if we don't rely on the threads being interruptible.

To monitor both threads at once, we can poll both futures from the main thread in a loop like this:
```java
Future<?> processCheck = execService.run(processCheckRunnable);
Future<?> rpcCheck = execService.run(rpcCheckRunnable);
while (!rpcCheck.isDone()) {
  if (processCheck.isDone()) {
    // fail because process failed, call processCheck.get() for the detailed exception
  }
  CommonUtils.sleepMs(Constants.SECOND_MS);
}
// check that rpcCheck succeeded and run one last process check, then return success
```
to make this work, `processCheckRunnable` needs to run a `while(true)` loop and wait for a second between each iteration.

Reply:
Got it, I will try this, thanks!

Reply:
Done